aten/src/ATen/native/cpu/FlashAttentionKernel.cpp

```diff
@@ -0,0 +1,516 @@
+            CblasRowMajor,
+            CblasNoTrans,
+            CblasNoTrans,
+            qBlockSize,
+            headSize,
+            kvBlockSize,
+            1.f,
+            qk_norm.data_ptr<scalar_t>() + ompIdx * qSplitSize * kvSplitSize,
+            kvBlockSize,
+            value + i * kvSize * vStride + headSize * j +
+                l * kvSplitSize * vStride,
+            vStride,
+            l == 0 ? 0.f : 1.f,
+            dst_fp32.data_ptr<float>() + ompIdx * qSplitSize * headSize,
+            headSize);
+      }
+      _reorder_mha_output_kernel<scalar_t>(
+          dst_fp32.data_ptr<float>() + ompIdx * qSplitSize * headSize,
+          output + i * qSize * hiddenSize +
+              headSize * j + k * qSplitSize * hiddenSize,
+          qBlockSize,
+          headSize,
+          hiddenSize);
+      data_index_step(i, batchSize, j, num_head, k, qSlice);
+    }
+  });
+
+}
+
+// Return the first output with other outputs arbitrary because they are used
+// for backward.
+// TODO: Calculate other outputs when adding cpu backward.
+// template <typename scalar_t>
+void cpu_flash_attention(
+    const Tensor& output,
+    const Tensor& logsumexp,
+    const Tensor& cum_seq_q,
+    const Tensor& cum_seq_k,
+    int64_t& max_q,
+    int64_t& max_k,
+    const Tensor& philox_seed,
+    const Tensor& philox_offset,
+    const Tensor& debug_attn_mask,
+    const at::Tensor& query,
+    const at::Tensor& key,
+    const at::Tensor& value,
+    double dropout_p,
+    bool is_causal,
+    bool return_debug_mask,
+    c10::optional<double> scale) {
+  // // Check validation
+  // TORCH_CHECK(at::hasMKL());
+  // TORCH_CHECK(query.scalar_type() == at::kFloat
+  //     || query.scalar_type() == at::kBFloat16);
+  // TORCH_CHECK(query.stride(1) == query.size(-1));
+  // TORCH_CHECK(query.dim() == 4);
+  // TORCH_CHECK(dropout_p == 0.0);
+  // TORCH_CHECK(!is_causal);
+
+  // Query (Batch x Num_heads x Q_seq_len  x Dim_per_head)
+  // Key   (Batch x Num_heads x KV_seq_len x Dim_per_head)
+  // Value (Batch x Num_heads x KV_seq_len x Dim_per_head)
+  int64_t batchSize = query.size(0);
+  int64_t qSize = query.size(2);
+  int64_t kvSize = value.size(2);
+  int64_t num_head = query.size(1);
+  int64_t headSize = query.size(3);
+  int64_t hiddenSize = num_head * headSize;
+
+  // Query -> Query(Batch x Q_seq_len x hiddenSize)
+  // Key   -> Key(Batch x KV_seq_len x hiddenSize)
+  // Value -> Value(Batch x KV_seq_len x hiddenSize)
+  Tensor q = query.transpose(1, 2).reshape(
+      {batchSize, qSize, hiddenSize});
+  Tensor k = key.transpose(1, 2).reshape(
+      {batchSize, kvSize, hiddenSize});
+  Tensor v = value.transpose(1, 2).reshape(
+      {batchSize, kvSize, hiddenSize});
+
+  int64_t qStride = q.stride(1);
+  int64_t kStride = k.stride(1);
+  int64_t vStride = v.stride(1);
+
+  const auto softmax_scale =
+      sdp::calculate_scale(query, scale).as_float_unchecked();
+
+  if (query.scalar_type() == at::kFloat) {
```

Comment:
the standard approach here is to use `AT_DISPATCH_FLOATING_TYPES_AND(ScalarType::BFloat16, query.scalar_type(),`